internal/moduletest/graph/transform_state.go

```diff
@@ -48,11 +62,48 @@ func (t *TestStateTransformer) Transform(g *terraform.Graph) error {
 	for node := range dag.SelectSeq(g.VerticesSeq(), runFilter) {
 		key := node.run.Config.StateKey
 		if _, exists := statesMap[key]; !exists {
-			state := &TestFileState{
-				File:  t.File,
-				Run:   nil,
-				State: states.NewState(),
+
+			var state *TestFileState
+
+			if bc, exists := t.File.Config.BackendConfigs[key]; exists {
+				// If the state for that state key should come from a backend,
+				// obtain and use that
+				if t.BackendFactory == nil {
+					return fmt.Errorf("error retrieving state for state key %q from backend: nil BackendFactory. This is a bug in Terraform and should be reported.", key)
+				}
+
+				f := t.BackendFactory(bc.Backend.Type)
+				if f == nil {
+					return fmt.Errorf("error retrieving state for state key %q from backend: No init function found for backend type %q. This is a bug in Terraform and should be reported.", key, bc.Backend.Type)
+				}
+				be, err := getBackendInstance(key, bc.Backend, f)
+				if err != nil {
+					return err
+				}
+
+				stmgr, err := be.StateMgr(backend.DefaultStateName) // We only allow use of the default workspace
+				if err != nil {
+					return fmt.Errorf("error retrieving state for state key %q from backend: error retrieving state manager: %w", key, err)
+				}
+
+				log.Printf("[TRACE] TestConfigTransformer.Transform: set initial state for state key %q using backend of type %T declared at %s", key, be, bc.Backend.DeclRange)
+				state = &TestFileState{
```

Comment:
While we are attaching the state here, our ctx.SetFileState ( in apply.go) overrides that state, and it does not seem to be setting `runBackend` at all. I think we would start running into panics there. 
Perhaps the ctx.SetFileState should be accepting all attributes that it want replaced rather than replacing the entire struct and therefore erroneously discarding things

Reply:
That's a good point about how the TestFileState.backend value is lost on subsequent updates to the internal state file.

Currently I think that is ok, as that `backend` value is only required for access at the point when it's still available, and there is a nil pointer check. However I think in future when we add the new cleanup behaviour we'd probably need to access the backend again, so your proposed change makes sense - I'll add it

Reply:
Thanks!. I agree it's fine to overwrite the exported field.